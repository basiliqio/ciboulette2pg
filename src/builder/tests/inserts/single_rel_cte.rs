use super::*;

#[test]
fn simple() {
    let mut builder = Ciboulette2PostgresBuilder::default();
    let store = gen_bag();
    let rel_table = Ciboulette2PostgresTableSettings::new(
        Ciboulette2PostgresId::Uuid(Ciboulette2PostgresSafeIdent::try_from("id").unwrap()),
        Some(Ciboulette2PostgresSafeIdent::try_from("public").unwrap()),
        Ciboulette2PostgresSafeIdent::try_from("favorite_color").unwrap(),
        store.get_type("favorite_color").unwrap(),
    );
    let main_cte_table = Ciboulette2PostgresTableSettings::new(
        Ciboulette2PostgresId::Uuid(Ciboulette2PostgresSafeIdent::try_from("id").unwrap()),
        Some(Ciboulette2PostgresSafeIdent::try_from("public").unwrap()),
        Ciboulette2PostgresSafeIdent::try_from("cte_peoples").unwrap(),
        store.get_type("peoples").unwrap(),
    );
    let url = Url::parse("http://localhost/peoples").unwrap();
    let ciboulette_request = gen_req_create_people(&store, &url);
    builder
        .gen_select_cte_single_rel(
            &rel_table,
            store.get_type("favorite_color").unwrap(),
            ciboulette_request.query(),
            &main_cte_table,
            &Ciboulette2PostgresSafeIdent::try_from("favorite_color").unwrap(),
        )
        .unwrap();

    let res = builder.build().unwrap();
    test_sql!(res);
}

#[test]
fn sparse() {
    let mut builder = Ciboulette2PostgresBuilder::default();
    let store = gen_bag();
    let rel_table = Ciboulette2PostgresTableSettings::new(
        Ciboulette2PostgresId::Uuid(Ciboulette2PostgresSafeIdent::try_from("id").unwrap()),
        Some(Ciboulette2PostgresSafeIdent::try_from("public").unwrap()),
        Ciboulette2PostgresSafeIdent::try_from("favorite_color").unwrap(),
        store.get_type("favorite_color").unwrap(),
    );
    let main_cte_table = Ciboulette2PostgresTableSettings::new(
        Ciboulette2PostgresId::Uuid(Ciboulette2PostgresSafeIdent::try_from("id").unwrap()),
        Some(Ciboulette2PostgresSafeIdent::try_from("public").unwrap()),
        Ciboulette2PostgresSafeIdent::try_from("cte_peoples").unwrap(),
        store.get_type("peoples").unwrap(),
    );
    let url = Url::parse("http://localhost/peoples?fields[favorite_color]=color").unwrap();
    let ciboulette_request = gen_req_create_people(&store, &url);
    builder
        .gen_select_cte_single_rel(
            &rel_table,
            store.get_type("favorite_color").unwrap(),
            ciboulette_request.query(),
            &main_cte_table,
            &Ciboulette2PostgresSafeIdent::try_from("favorite_color").unwrap(),
        )
        .unwrap();

    let res = builder.build().unwrap();
    test_sql!(res);
}

#[test]
fn empty_sparse() {
    let mut builder = Ciboulette2PostgresBuilder::default();
    let store = gen_bag();
    let rel_table = Ciboulette2PostgresTableSettings::new(
        Ciboulette2PostgresId::Uuid(Ciboulette2PostgresSafeIdent::try_from("id").unwrap()),
        Some(Ciboulette2PostgresSafeIdent::try_from("public").unwrap()),
        Ciboulette2PostgresSafeIdent::try_from("favorite_color").unwrap(),
        store.get_type("favorite_color").unwrap(),
    );
    let main_cte_table = Ciboulette2PostgresTableSettings::new(
        Ciboulette2PostgresId::Uuid(Ciboulette2PostgresSafeIdent::try_from("id").unwrap()),
        Some(Ciboulette2PostgresSafeIdent::try_from("public").unwrap()),
        Ciboulette2PostgresSafeIdent::try_from("cte_peoples").unwrap(),
        store.get_type("peoples").unwrap(),
    );
    let url = Url::parse("http://localhost/peoples?fields[favorite_color]=").unwrap();
    let ciboulette_request = gen_req_create_people(&store, &url);
    builder
        .gen_select_cte_single_rel(
            &rel_table,
            store.get_type("favorite_color").unwrap(),
            ciboulette_request.query(),
            &main_cte_table,
            &Ciboulette2PostgresSafeIdent::try_from("favorite_color").unwrap(),
        )
        .unwrap();

    let res = builder.build().unwrap();
    test_sql!(res);
}
