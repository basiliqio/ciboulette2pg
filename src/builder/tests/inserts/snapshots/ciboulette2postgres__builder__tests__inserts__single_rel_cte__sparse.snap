---
source: src/builder/tests/inserts/single_rel_cte.rs
expression: "sqlformat::format(query.as_str(), &sqlformat::QueryParams::None,\n                  sqlformat::FormatOptions::default())"

---
SELECT
  "public"."favorite_color"."id" :: TEXT AS "id",
  "public"."favorite_color"."id" AS "main_id",
  $1 :: TEXT AS "type",
  JSON_BUILD_OBJECT($2, "public"."favorite_color"."color") AS "data",
  "public"."cte_peoples"."id" :: TEXT AS "related_id",
  $3 :: TEXT AS "related_type"
FROM
  "public"."favorite_color"
  INNER JOIN "public"."cte_peoples" ON "public"."favorite_color"."id" = "public"."cte_peoples"."favorite_color";
